patroni/dcs/__init__.py

```diff
@@ -778,16 +778,69 @@ def from_node(version: _Version, value: str) -> 'TimelineHistory':
         return TimelineHistory(version, value, lines)
 
 
+class Status(NamedTuple):
+    """Immutable object (namedtuple) which represents `/status` key.
+
+    Consists of the following fields:
+
+    :ivar last_lsn: :class:int object containing position of last known leader LSN.
+    :ivar slots: state of permanent logical replication slots on the primary in the format: {"slot_name": int}.
+    """
+    last_lsn: int
+    slots: Optional[Dict[str, int]]
+
+    @staticmethod
+    def empty() -> 'Status':
+        """Construct an empty :class:`Status` instance.
+
+        :returns: empty :class:`Status` object.
+        """
+        return Status(0, None)
+
+    @staticmethod
+    def from_node(value: Union[str, Dict[str, Any], None]) -> 'Status':
+        """Factory method to parse *value* as :class:`Status` object.
+
+        :param value: JSON serialized string
+
+        :returns: constructed :class:`Status` object.
+        """
+        slots: Union[str, Dict[str, int], None] = None
+        last_lsn = 0
+        try:
+            if isinstance(value, str):
+                value = json.loads(value)
+            if isinstance(value, int):  # legacy
+                return Status(value, None)
+            assert isinstance(value, dict)
+            last_lsn = value.get('optime')
+            slots = value.get('slots')
+        except Exception:
```

Comment:
I don't think there is big value in logging errors here.
If there is something unexpected the primary will overwrite it on the next heartbeat.

Reply:
Unless there is a persistent problem that causes this to throw an exception every time and then a broad exception hides what is happening. Do we know that this is definitely JSON and not some service error, or crud from something entirely unexpected? Is that caught and recorded somewhere else, perhaps in the specific DCS implementation or in the underlying service module we're using? If it is and we know the input has been sanitised then why do we need an `Exception` clause? A debug level log could be good in case we want to hide it in normal circumstances but elevate the log level when the situation is not clear. Could we be hiding potentially useful information from someone trying to debug an issue?

Reply:
If the primary fails to update the `/status` key (or the leader ConfigMap/Endpoint in case of Kubernetes) there will be a clear error.

Reply:
Ok, the inverse of that should be true then yes? As this namedtuple is populated on read of the service. That will depend on whether the value is JSON and the structure is sane right? e.g. in etc DCS

in `_cluster_from_nodes`
```
        status = nodes.get(self._STATUS) or nodes.get(self._LEADER_OPTIME)
        status = Status.from_node(status and status['Value'])
```

and caller (one of them)
``` 
    def _cluster_loader(self, path: str) -> Cluster:
        result = self.retry(self._client.read, path, recursive=True, quorum=self._ctl)
        nodes = {node.key[len(result.key):].lstrip('/'): node for node in result.leaves}
        return self._cluster_from_nodes(result.etcd_index, nodes)
```

So we don't make sure that the `status['Value']` is JSON before it's passed to `Status.from_node` but we should see an error if the response from etcd (or intermediate service) is malformed. And what you're saying is that `status['Value']` should be reset later if it is. This goes back to the thing you said about ensuring patroni is resilient to external interference.

I think I get it :wink:

Reply:
Do we test for that?

Reply:
Test for what?

Reply:
I guess it could be cool having a debug message in case the input is malformed.

Reply:
Also, I'd say to catch `json.decoder.JSONDecodeError` instead of a general `Exception`. The former is what I expect to see if we have issues in decoding the JSON object.

Reply:
Uncaught exception may do a lot of damage.
I want to be on the safe side and therefore catching everything.

Reply:
Ok, your call :)